Adds conversion comment refresh requests via Private Messages

shrink opened the pull request:
This is my first attempt at implementing refresh requests via Private Messages sent to the bot. I am not very familiar with nodejs so I hit a stumbling block with writing tests, — any guidance on that would be greatly appreciated. Specifically, the function `privateMessageFunction` is [being called at the start of each Private Message test](https://github.com/cannawen/metric_units_reddit_bot/blob/master/test/bot-test.js#L282) but I can’t find what that function does — [where does `f` come from?](https://github.com/cannawen/metric_units_reddit_bot/blob/master/test/bot-test.js#L83)

There are quite a few erroneous tab/space characters in the `bot.js` and `network.js` files which my editor (Atom) automatically removed. I couldn't find a way to stop Atom removing them, so my commit is polluted with those changes. Sorry :-( 

Left to do before this is mergeable:

- [x] Add tests
- [x] Add analytics
- [x] Add the footer link for creating a PM

Some notes on my implementation:

- Requires 3 API requests to complete the refresh process.
- Refreshes when a PM containing "refresh {id}" as the subject is received.
- `postTitle` is a required value for the conversion process, however, the comment endpoint does not return the value so I have used `This is a dummy value.` to allow it to function instead of refactoring the comment conversion process as to keep my changes small.
- Does not reply to the user to confirm the refresh request has been received — I'm not sure if Private Messages are rate limited by reddit, can add this if there aren't concerns about hitting Private Message limits.
- If the ID provided does not return a value from the reddit API then the refresh request is skipped.
- If the comment no longer has anything to convert then no edit is made.

When a comment has been edited and no longer contains any values that need to be converted, we could either delete the bot reply or we could edit it. I did not implement deleting the reply because I think it might be fun if instead the bot reply is edited with personality, something referencing the fact that the bot was made to do work unnecessarily. This would be a fairly easy feature to implement and would be high impact so I think it would be good to leave it as something for a first timer to implement instead of implementing it as part of this change. 

Closes #67

shrink committed:
Adds conversion comment refresh requests via Private Messages

This is my first attempt at implementing refresh requests via Private
Messages sent to the bot. I am not very familiar with nodejs so I hit a
stumbling block with writing tests, — any guidance on that would be
greatly appreciated. Specifically, the function
`privateMessageFunction` is being called at the start of each Private
Message test but I can’t find what that function does.

- [ ] Add tests
- [ ] Add analytics

Some notes on my implementation:

- Requires 3 API requests to complete the refresh process
- `postTitle` is a required value for the conversion process, however
the comment endpoint does not return the value so I have used `This is
a dummy value.` to allow it to function instead of refactoring the
comment conversion process as to keep my changes small
- Does not reply to the user to confirm the refresh request
- If the ID provided does not return a value from the reddit API then
the refresh request is skipped
- If the comment no longer has anything to convert then no edit is made

If a comment is edited and no longer contains any values that need
converting the comment the bot made could be deleted, or maybe instead
the comment could be edited to have a sassy personality reply,
something referencing the fact that the commenter caused the bot to do
some work and now that work was in vain? I didn’t implement this
because I think it would be a good feature to allow a first timer to
implement because they’d get to have a big impact on the way the bot
acts.

chazzlabs commented:
Hey @citricsquid!

>Specifically, the function privateMessageFunction is being called at the start of each Private Message test but I can’t find what that function does — where does f come from?

Take a look [here](https://github.com/cannawen/metric_units_reddit_bot/blob/master/test/bot-test.js#L78).  On that line we're creating a new function and assigning it to `helperStub.setIntervalSafely`.  The first argument of that function is `f`, and we're assigning that value to either `commentFunction` or `privateMessageFunction`.  `helperStub` is a stubbed version of the [`helper` module](https://github.com/cannawen/metric_units_reddit_bot/blob/master/src/helper.js) which helps facilitate unit testing.  You can see in the `helper` module that the [`setIntervalSafely`](https://github.com/cannawen/metric_units_reddit_bot/blob/master/src/helper.js#L25) function is calling Javascript's native `setInterval` function.  This function calls its first argument, a callback function, every 'x' milliseconds, where 'x' is the second argument passed into the function (in this case, `seconds` * 1000).  

Stubbing out this function in our unit test means that the `setInerval` function is not actually called; instead our stub is called.  This is useful in this case because we don't want our unit tests executing code that may 1) delay the execution of the subsequent test cases/suites or 2) cause the unit test's assertions to be run before the callback in the `setInterval` function, thus causing tests to fail.  Because our stub is called, this allows us to save the `f` function received via the first argument and call it explicitly to facilitate the execution of our test cases.

Ultimately, the answer to your question, "Where does `f` come from?" is: [lines 31 and 32 of the `helper` module](https://github.com/cannawen/metric_units_reddit_bot/blob/master/src/bot.js#L31).

I apologize if this answer is much more detailed than you might need, but I wanted to be thorough in case anyone else reading this might benefit from it.  Please let me know if you have questions.

shrink commented:
@chazzlabs that's a perfect level of detail :) thank you! I'll work on adding some tests now.

chazzlabs commented:
@citricsquid Awesome!  I'm glad that was helpful.  If you have any other questions please don't hesitate to post them here.

shrink committed:
Refactors network requests to return formatted data

Previously returned the raw API response which then had to be formatted
each time it was used, now they return the data formatted ready for use.

shrink committed:
Adds tests for refresh requests

shrink commented:
I have refactored the network code for getting comments and their replies. I have added 3 tests for the refresh request functionality. I'm not super confident in the quality of the tests but I *think* they're in the right direction. Any pointers for improvements would be great! Thank you.

shrink committed:
Adds refresh conversion link to comment footer, and refactors footer compilation

shrink commented:
I have added the `refresh conversion` link to the footer of the comments and refactored the process of compiling the comment footer. Each item in the footer is now managed through an array of footer items, and there is a function for generating reddit compose message links which makes use of the `stopMessage` value.

shrink commented on src/reply_maker.js:
```diff
@@ -17,22 +18,31 @@ function formatReply(comment, conversions) {
 
+  const items = [
+    {"value" : "metric units " + species },
+    {"type" : "link", "value" : "feedback", "href" : "https://redd.it/73edn2" },
```
Short links reduce message size (in bytes, which robots totally care about) and make the code easier to read.

shrink committed:
Adds analytics for successful conversion updates

cannawen commented on src/bot.js:
```diff
@@ -132,6 +132,31 @@ function replyToMessages() {
+      const commentId = message['subject'].match(/refresh (\w+)/i)[1];
+      const comment = network.getComment('t1_' + commentId);
+
+      if (! comment) return;
```
This is really just a style thing, but I prefer if statements always being written like
```
if (! comment) {
  return;
}
```
I feel like always having the curly braces & always adding an endline (even on a one-line if) makes it more obvious where the if statement starts and ends, and this scoping prevents bugs. Open to discussion though

shrink replied:
Sure, I agree that there are advantages to being explicit with short statements. I will change the style of my return statements.

cannawen commented on src/network.js:
```diff
@@ -171,6 +171,36 @@ function postComment(parentId, markdownBody) {
+    'author': data['author'],
+    'id': data['name'],
+    'link_id' : data['link_id'],
+    'postTitle': 'This is a dummy value.',
```
`This is a dummy value.`? Does it not return the post title?

shrink replied:
The reddit API is, unfortunately, inconsistent with the values it returns. The comment end point does not return a value for `postTitle`, but the project requires a `postTitle` value elsewhere so I added the dummy value. 

I didn't think it was worth an additional API request to get the `postTitle` as it does not (as far as I can tell) matter when updating an existing comment -- since for a comment to exist it has already passed `postTitle` validation. I can add an additional request to fetch that value though if it's needed.

https://www.reddit.com/api/info.json?id=t1_do2yo6a

cannawen replied:
I agree, no need to make an extra network call for this. 

Maybe we can change it to an empty string and add a comment beside it "Not returned by reddit endpoint"? When I first read it, I thought it might have been some accidental test code left in

shrink replied:
Very good point, done! :-)

shrink committed:
Adds trackEdit helper method for testing

cannawen commented on src/bot.js:
```diff
@@ -132,6 +132,31 @@ function replyToMessages() {
+
+      if (! comment) return;
+
+      const commentReplies = network.getCommentReplies(comment['link_id'], commentId);
```
An architecture change idea:

Instead of sending a message with the OP's message id and then searching through the comment replies, we could send a message with the bot's message id and it should have a free `parent_id` property attached to it (...I think). That way we don't need the function `getCommentReplies`, we can just reuse `getComment`

shrink replied:
I tried this originally and unfortunately got all the way through my implementation before I realised why it isn't suitable: to provide the `refresh conversion` link in the footer of the comment we need to include an ID, and the only ID available to us when replying is the ID of the parent comment, we don't have the bot reply ID until after the bot reply has been made.

There were 2 options: either edit every bot comment immediately after it is made to add the reply ID to the refresh request link, or use the comment replies to identify the bot reply when a refresh request is received.

I chose to go with the second method because it will require fewer API requests and reduces the surface of my changes. The first method would double the API requests the bot does for every conversion. I am open to switching to the edit every conversion comment method though if that's preferable :)

cannawen replied:
LOL, that's a tricky one (but totally makes sense now that you spell it out)! I think you made the right choice, getting child replies sound easier than editing every single comment.

cannawen commented on src/reply_maker.js:
```diff
@@ -17,22 +18,31 @@ function formatReply(comment, conversions) {
+    item.value = '^' + item.value.replace(/ /g, ' ^');
+    if (item.type == 'link') item.value = `[${item.value}](${item.href})`;
+    return item.value;
+  }).join(' ^| ');
```
This refactor looks awesome!!!!! Much easier to read and understand what is going on

cannawen commented on test/reply_maker-test.js:
```diff
@@ -40,5 +40,17 @@ describe('reply_maker', () => {
+      .include("[^block](https://www.reddit.com/message/compose")
+      .and
+      .include("[^refresh ^conversion](https://www.reddit.com/message/compose")
+    });
```
Thanks for adding these missing tests :D

cannawen reviewed (changes requested):
Hey, overall very good PR, I like how functionally everything is written!

One suggestion about a possible refactor of how the network calls are handled.

shrink committed:
Restyles single line statements into multi-line statements

cannawen commented:
I think it would be hilarious if the bot sassed people who edited their comments. 
@citricsquid can you open a separate issue for that feature request? XD

shrink committed:
Replaces dummy value with explanation of why postTitle is included

cannawen reviewed (approved):
Amazing! I'm really impressed with the quality of the PR, and how fast you finished it :D 

Just one last thing to confirm: If you do not want to add yourself to the all-contributors square you can merge this PR in now. 

If you would like to be added, follow the [instructions here](https://github.com/cannawen/metric_units_reddit_bot/blob/master/CONTRIBUTING.md#add-yourself-to-the-contributors-list) and merge in afterwards :)

shrink commented:
Thank you very much :-) I had fun! This is a great project, I like how well managed it is. Very much in the open source spirit.

I have added a feature request issue for the bot to sass users, #90. If I find time tomorrow I will go through and add some instructions (as you did in #79) so that a first timer can take a shot at implementing it.

I don't need to be added to the contributors.

nalinbhardwaj commented:
Interestingly, after this commit, it appears tests on my computer are failing. There is no apparent reason(no diff from cannawen/master), but I'm getting the following error:
![screen shot 2017-10-11 at 7 31 03 am](https://user-images.githubusercontent.com/6984346/31419016-fc7babc0-ae56-11e7-94d8-d85df70aa698.png)

Repeated 5 times across tests. Using node v8.6.0 and npm 5.4.2(both are >= the travis CI) on macOS.

shrink commented:
@nalinbhardwaj I am also running macOS (High Sierra, version 10.13 (17A405)) — with node v8.6.0, npm v5.4.2 — and the tests pass for me:

```
  reply_maker
    #formatReply()
      ✓ should create a tabular response for 1 conversion
      ✓ should create tabular response for 2 conversions
      ✓ should create tabular response of supersets of text
      ✓ should say metric units bot on normal subreddits
      ✓ should say metric units human on totallynotrobots
      ✓ should include footer links
```

I am unsure what the problem could be, as looking at that line (33 of reply_maker.js) I can't think why `item.value` would not exist: maybe your update of the repository contents failed in some way?
